runtime/compiler/optimizer/InlinerTempForJ9.cpp

```diff
@@ -895,11 +897,20 @@ TR_J9InlinerPolicy::genCodeForUnsafeGetPut(TR::Node* unsafeAddress,
    TR::TreeTop* nullComparisonTree = nullComparisonBlock->getLastRealTreeTop();
    TR::TreeTop *nullComparisonEntryTree = nullComparisonBlock->getEntry();
    TR::TreeTop *nullComparisonExitTree = nullComparisonBlock->getExit();
-   //if conversionNeeded is true, we haven't generated and we don't need arrayDirectAccessBlock
-   TR::Block *arrayDirectAccessBlock = conversionNeeded ? nullComparisonTree->getNode()->getBranchDestination()->getNode()->getBlock() : NULL;
-   TR::Block *indirectAccessBlock;
+
+   //if arrayBlockNeeded is false, we haven't generated and we don't need arrayDirectAccessBlock
+   TR::Block *arrayDirectAccessBlock = arrayBlockNeeded ? nullComparisonTree->getNode()->getBranchDestination()->getNode()->getBlock() : NULL;
    TR::Block * directAccessBlock;
-   if (conversionNeeded)
+   TR::Block *indirectAccessBlock;
+
+   // There are 5 possible cases determining which blocks should be added to the generated IL trees
+   // 1.) checksNeeded && arrayCheckNeeded && arrayBlockNeeded: add directAccess, indirectAccess, and arrayDirectAccess
```

Comment:
for non-offheap policy (i.e.  normal object and array headers are of same size), we can optimize the cases down to the fact that we don't need arrayCheck, since object and array accesses can use the same tree. i presumed you took advantage of that optimization in higher level of routines where you can always set arrayCheckNeeded to false (for gencon e.g.).

Reply:
In the case where the data being read/written is less than 4 bytes, object and array accesses are actually handled differently, as outlined [here](https://github.com/eclipse-openj9/openj9/blob/671e76243d8e011486e26f4d2d298012efe2e2d9/runtime/compiler/optimizer/InlinerTempForJ9.cpp#L1041-L1111): 
```
Converting Unsafe.get/put* routines into inlined code involves two cases:
1) if the size of the element to put/get is 4 bytes or more
2) if the size of the element to put/get is less than 4 bytes (boolean, byte, char, short)
.
.
.
In (2), there are three alternatives on how to read from/write the object. direct,
direct with conversion, indirect. The same three conditions are used to decide which one
to use based on the following pseudocode:
if (object is NULL)
  use direct access with conversion
else if (object is array)
  use direct access with conversion
else if (object is of type Class)
  use indirect access
else
  use direct access
```
...which is why I still included the array check in the case where gencon GC policy is being used and the object type is unknown at compile time.